drivers/sdhc/rcar_mmc.c

```diff
@@ -0,0 +1,2388 @@
+	return 0;
+}
+
+static const struct sdhc_driver_api rcar_sdhc_api = {
+	.card_busy = rcar_mmc_card_busy,
+#ifdef CONFIG_RCAR_MMC_SCC_SUPPORT
+	.execute_tuning = rcar_mmc_execute_tuning,
+#endif
+	.get_card_present = rcar_mmc_get_card_present,
+	.get_host_props = rcar_mmc_get_host_props,
+	.request = rcar_mmc_request,
+	.reset = rcar_mmc_reset,
+	.set_io = rcar_mmc_set_io,
+};
+
+/**
+ * @brief Start SD-IF clock at max frequency configured in dts
+ *
+ * @param cfg The Renesas MMC driver configuration
+ *
+ * @retval 0 on success, negative errno code otherwise
+ */
+static int rcar_mmc_init_start_clk(const struct mmc_rcar_cfg *cfg)
+{
+	int ret = 0;
+	const struct device *cpg_dev = cfg->cpg_dev;
+	uintptr_t rate = cfg->max_frequency;
+
+	ret = clock_control_on(cpg_dev, (clock_control_subsys_t *)&cfg->bus_clk);
+	if (ret < 0) {
+		return ret;
+	}
+
+	ret = clock_control_on(cpg_dev, (clock_control_subsys_t *)&cfg->cpg_clk);
+	if (ret < 0) {
+		return ret;
+	}
+
+	ret = clock_control_set_rate(cpg_dev, (clock_control_subsys_t *)&cfg->cpg_clk,
+				     (clock_control_subsys_rate_t)rate);
+	if (ret < 0) {
+		clock_control_off(cpg_dev, (clock_control_subsys_t *)&cfg->cpg_clk);
+	}
+
+	rate = MMC_BUS_CLOCK_FREQ;
+	ret = clock_control_set_rate(cpg_dev, (clock_control_subsys_t *)&cfg->bus_clk,
+				     (clock_control_subsys_rate_t)rate);
+	/* SD spec recommends at least 1 ms of delay after start of clock */
+	k_msleep(1);
+
+	return ret;
+}
+
+/**
+ * @brief Initialize Host properties for working with SDHC and MMC stack
+ *
+ * @param dev The pointer to an MMC device structure
+ *
+ * @retval none
+ */
+static void rcar_mmc_init_host_props(const struct device *dev)
+{
+	struct mmc_rcar_data *data = dev->data;
+	const struct mmc_rcar_cfg *cfg = dev->config;
+	struct sdhc_host_props *props = &data->props;
+	struct sdhc_host_caps *host_caps = &props->host_caps;
+
+	memset(props, 0, sizeof(*props));
+
+	/* Note: init only properties that are used for mmc/sdhc */
+
+	props->f_max = cfg->max_frequency + MMC_MAX_FREQ_CORRECTION;
+	/*
+	 * note: actually, it's possible to get lower frequency
+	 *       if we use divider from cpg too
+	 */
+	props->f_min = (cfg->max_frequency >> 9);
+
+	props->power_delay = 100; /* ms */
+
+	props->is_spi = 0;
+
+	switch (cfg->bus_width) {
+	case SDHC_BUS_WIDTH8BIT:
+		host_caps->bus_8_bit_support = 1;
+	case SDHC_BUS_WIDTH4BIT:
+		host_caps->bus_4_bit_support = 1;
+	default:
+		break;
+	}
+
+	host_caps->high_spd_support = 1;
+#ifdef CONFIG_RCAR_MMC_SCC_SUPPORT
+	host_caps->sdr104_support = cfg->mmc_sdr104_support;
+	host_caps->sdr50_support = cfg->uhs_support;
+	/* neither Linux nor U-boot support DDR50 mode, that's why we don't support it too */
+	host_caps->ddr50_support = 0;
+	host_caps->hs200_support = cfg->mmc_hs200_1_8v;
+	/* TODO: add support */
+	host_caps->hs400_support = 0;
+#endif
+
+	host_caps->vol_330_support =
+		regulator_is_supported_voltage(cfg->regulator_vqmmc, 3300000, 3300000);
```

Comment:
I'm confused what the regulator driver is used for here. It seems these fixed regulators cannot be enabled or disabled, and are just being used to determine the voltage ranges supported by the SDHC? Or is there a method to enable and disable these regulators I'm missing?